src/blockly/generators/propc/gpio.js

```diff
@@ -1740,36 +1728,58 @@ Blockly.Blocks.sd_read = {
         this.setSdMode(mode);
     },
     setSdMode: function (mode) {
-        if(this.getInput('SIZE')) {
+        if (this.getInput('SIZE')) {
+            var valueConnection = this.getInput('SIZE').connection;
+            if (valueConnection) {
+                var connectedBlock = valueConnection.targetBlock()
+            }
             this.removeInput('SIZE');
         }
         if (this.getInput("VALUE")) {
             this.removeInput("VALUE");
         }
         if (mode === "fwrite") {
-            this.appendValueInput("SIZE");
+            this.appendValueInput("SIZE")
+                    .setCheck(null)
+                    .appendField("SD file")
+                    .appendField(new Blockly.FieldDropdown([
+                        ["write", "fwrite"],
+                        ["read", "fread"],
+                        ["close", "fclose"]
+                    ], function (mode) {
+                        this.sourceBlock_.setSdMode(mode);
+                    }), "MODE");
             this.appendValueInput("VALUE")
                     .setCheck("String")
                     .appendField("bytes of");
         } else if (mode === "fread") {
             this.appendValueInput("SIZE")
-                    .setCheck("Number");
+                    .setCheck("Number")
+                    .appendField("SD file")
+                    .appendField(new Blockly.FieldDropdown([
+                        ["read", "fread"],
+                        ["write", "fwrite"],
+                        ["close", "fclose"]
+                    ], function (mode) {
+                        this.sourceBlock_.setSdMode(mode);
+                    }), "MODE");
             this.appendDummyInput("VALUE")
                     .appendField("bytes  store in")
                     .appendField(new Blockly.FieldVariable(Blockly.LANG_VARIABLES_SET_ITEM), 'VAR');
         } else {
-            this.appendDummyInput("SIZE");
+            this.appendDummyInput("SIZE")
+                    .appendField("SD file")
+                    .appendField(new Blockly.FieldDropdown([
+                        ["close", "fclose"],
+                        ["read", "fread"],
+                        ["write", "fwrite"]
+                    ], function (mode) {
+                        this.sourceBlock_.setSdMode(mode);
+                    }), "MODE");
```

Comment:
These additions address the issue - the desired menu option becomes the first option in the menu, which makes it so that the menu's value doesn't have to be set later, which would then trigger the menu's change handler and start the infinite loop.